### ISSUE TITLE
Cannot resolve forward reference Callable

### ISSUE THREAD

**Issue by mikkelee** [tox-dev/sphinx-autodoc-typehints#76]
I'm working on documenting some code and ran into a problem with the below code:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, TYPE_CHECKING

@dataclass
class Bin:
    description: str
    matcher: Callable[..., bool]
```

Docs are successfully generated, but the below warning is also output:

```console
WARNING: Cannot resolve forward reference in type annotations of "Bin": name 'Callable' is not defined
```

I use future annotations and TYPE_CHECKING elsewhere, don't know if that is relevant but I included them here to avoid potential confusion.

**Comment by rmorshea** [comment]
I'm experiencing this as well.

**Comment by rmorshea** [comment]
This might become a bigger problem when all type hints will be "forward references":

https://www.python.org/dev/peps/pep-0563/

**Comment by agronholm** [comment]
I'm confused – sphinx-autodoc-typehints does not even currently support variable annotations so it should not be trying to resolve them. Can you dig a bit deeper please?

**Comment by rmorshea** [comment]
Actually I didn't catch the use of `__futures__` here, but I am getting warnings about forward references to variables. I'll see if I can dig deeper at some point, but from what I can tell right now, in my case it's trying to trace forward references to variables that are used in function definitions.

**Comment by blueyed** [comment]
I've tried sphinx-autodoc-typehints with pytest's docs (https://github.com/pytest-dev/pytest/pull/6665), and with that it seems like it triggers those warnings then with circular imports.
(I assume types in quotes should work in general?)

**Comment by adamtheturtle** [comment]
I believe that this may be https://bugs.python.org/issue34776 and therefore fixed in newer versions of Python (3.8.1+, and I think 3.7.2+).

**Comment by gaborbernat** [comment]
Closing per above comment.
